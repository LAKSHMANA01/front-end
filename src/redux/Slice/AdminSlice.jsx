import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

// Fetch all tasks
export const fetchAllTasks = createAsyncThunk('admin/tasks/fetchAll', async (_, { rejectWithValue }) => {
  try {
    const response = await axios.get('https://localhost:8000/api/admin/tasks'); // Replace with your API
    return response.data;
  } catch (error) {
    return rejectWithValue(error.response?.data || "Failed to fetch tasks");
  }
});

// Fetch all users
export const fetchAllUsers = createAsyncThunk('admin/users/fetchAll', async (_, { rejectWithValue }) => {
  try {
    const response = await axios.get('https://localhost:8000/api/admin/users'); // Replace with your API
    return response.data;
  } catch (error) {
    return rejectWithValue(error.response?.data || "Failed to fetch users");
  }
});

export const fetchAllEngineers = createAsyncThunk("admin/engineers/fetchAll",async (_, { rejectWithValue }) => {
    try {
      const response = await axios.get("https://localhost:8000/api/admin/engineers"); // Replace with your API
      return response.data;
    } catch (error) {
      return rejectWithValue(error.response?.data || "Failed to fetch engineers");
    }
  }
);

export const fetchDeferredTasks = createAsyncThunk(
  "admin/deferredTasks/fetchAll",
  async (_, { rejectWithValue }) => {
    try {
      const response = await axios.get("https://localhost:8000/api/admin/status/deferred"); // Replace with your API
      return response.data;
    } catch (error) {
      return rejectWithValue(error.response?.data || "Failed to fetch deferred tasks");
    }
  }
);

// Fetch available engineers
export const fetchAvailableEngineers = createAsyncThunk(
  'tasks/fetchAvailableEngineers',
  async () => {
    try {
      const response = await axios.get('/api/engineers/available');
      return response.data.engineers; // Assuming API returns an array of engineers
    } catch (error) {
      throw new Error(error.response?.data?.message || error.message); // Handle error
    }
  }
);

// Reassign ticket to an engineer
export const reassignTicket = createAsyncThunk(
  'tasks/reassignTicket',
  async ({ ticketId, engineerId }) => {
    try {
      const response = await axios.patch(`/api/reassign/${ticketId}/${engineerId}`, {}, {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      return response.data; // Assuming the response has details of reassigned ticket
    } catch (error) {
      throw new Error(error.response?.data?.message || error.message); // Handle error
    }
  }
);

export const fetchEngineerTasks = createAsyncThunk(
  "admin/fetchEngineerTasks",
  async (engineerId, { rejectWithValue }) => {
    try {
      console.log("engineerId", engineerId);
      const response = await axios.get(`https://localhost:8000/api/tasks/engineer/${engineerId}`); // âœ… API call
      return response.data;
    } catch (error) {
      return rejectWithValue(error.response?.data || "Error fetching tasks");
    }
  }
);



const adminSlice = createSlice({
  name: 'admin',
  initialState: {
    tasks: [],
    completedTasks: [],
    deferredTasks: [],
    users: [],
    engineers: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchAllTasks.pending, (state) => { state.loading = true; })
      .addCase(fetchAllTasks.fulfilled, (state, action) => {
        state.tasks = action.payload;
        state.loading = false;
      })
      .addCase(fetchAllTasks.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      .addCase(fetchAllUsers.pending, (state) => { state.loading = true; })
      .addCase(fetchAllUsers.fulfilled, (state, action) => {
        state.users = action.payload;
        state.loading = false;
      })
      .addCase(fetchAllUsers.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      .addCase(fetchAllEngineers.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchAllEngineers.fulfilled, (state, action) => {
        state.engineers = action.payload;
        state.loading = false;
      })
      .addCase(fetchAllEngineers.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      .addCase(fetchDeferredTasks.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchDeferredTasks.fulfilled, (state, action) => {
        state.deferredTasks = action.payload;
        state.loading = false;
      })
      .addCase(fetchDeferredTasks.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      .addCase(fetchEngineerTasks.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchEngineerTasks.fulfilled, (state, action) => {
        state.loading = false;
        state.tasks = action.payload;
      })
      .addCase(fetchEngineerTasks.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })
      .addCase(fetchAvailableEngineers.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchAvailableEngineers.fulfilled, (state, action) => {
        state.loading = false;
        state.availableEngineers = action.payload;
      })
      .addCase(fetchAvailableEngineers.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
      
  },
});

export default adminSlice.reducer;
