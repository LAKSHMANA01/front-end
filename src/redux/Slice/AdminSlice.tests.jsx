import { configureStore } from "@reduxjs/toolkit";
import adminReducer, {
  fetchAllTasks,
  fetchAllUsers,
  fetchAllApprovedEngineers,
  fetchAllEngineers,
  approveEngineer,
  fetchDeferredTasks,
  fetchAvailableEngineers,
  reassignTicket,
  fetchEngineerTasks,
} from "./AdminSlice";
import apiClient from "../../utils/apiClient";


jest.mock("../../utils/apiClient");

describe("Admin Slice Async Thunks", () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: { admin: adminReducer }
    });
  });

  test("fetchAllTasks should fetch tasks and update state", async () => {
    const mockTasks = [{ id: 1, name: "Task 1" }];
    apiClient.get.mockResolvedValueOnce({ data: mockTasks });

    await store.dispatch(fetchAllTasks());
    expect(store.getState().admin.tasks).toEqual(mockTasks);
  });

  test("fetchAllTasks should handle error when fetching tasks fails", async () => {
    const errorResponse = { response: { data: "Failed to fetch tasks" } };
    apiClient.get.mockRejectedValueOnce(errorResponse);
    await store.dispatch(fetchAllTasks());
    expect(store.getState().admin.error).toEqual("Failed to fetch tasks");
  });

  test("fetchAllUsers should fetch users and update state", async () => {
    const mockUsers = [{ id: 1, name: "User 1" }];
    apiClient.get.mockResolvedValueOnce({ data: mockUsers });

    await store.dispatch(fetchAllUsers());
    expect(store.getState().admin.users).toEqual(mockUsers);
  });

  test("fetchAllUsers should handle error when fetching users fails", async () => {
    const errorResponse = { response: { data: "Failed to fetch users" } };
    apiClient.get.mockRejectedValueOnce(errorResponse);
    await store.dispatch(fetchAllUsers());
    expect(store.getState().admin.error).toEqual("Failed to fetch users");
  });

  test("fetchAllApprovedEngineers should update state with engineers", async () => {
    const mockEngineers = [{ email: "engineer@test.com", approved: true }];
    apiClient.get.mockResolvedValueOnce({ data: mockEngineers });

    await store.dispatch(fetchAllApprovedEngineers());
    expect(store.getState().admin.approvedEngineers).toEqual(mockEngineers);
  });

  test("fetchAllApprovedEngineers should handle error when fetching engineers fails", async () => {
    const errorResponse = { response: { data: "Failed to fetch engineers" } };
    apiClient.get.mockRejectedValueOnce(errorResponse);
    await store.dispatch(fetchAllApprovedEngineers());
    expect(store.getState().admin.error).toEqual("Failed to fetch engineers");
  });

  test("fetchAllEngineers should update state with engineers", async () => {
    const mockEngineers = [{ email: "engineer@test.com", approved: false }];
    apiClient.get.mockResolvedValueOnce({ data: { approvalEngineers: mockEngineers } });

    await store.dispatch(fetchAllEngineers());
    expect(store.getState().admin.engineers).toEqual(mockEngineers);
  });

  test("fetchAllEngineers should handle error when fetching engineers fails", async () => {
    const errorResponse = { response: { data: "Failed to fetch engineers" } };
    apiClient.get.mockRejectedValueOnce(errorResponse);
    await store.dispatch(fetchAllEngineers());
    expect(store.getState().admin.error).toEqual("Failed to fetch engineers");
  });

  test("approveEngineer should update state correctly", async () => {
    const mockEngineer = { email: "engineer@test.com", approved: true };
    apiClient.patch.mockResolvedValueOnce({ data: { engineer: mockEngineer } });

    await store.dispatch(approveEngineer({ engineerEmail: "engineer@test.com", approve: true }));
    expect(store.getState().admin.approvedEngineers).toContainEqual(mockEngineer);
  });

  test("approveEngineer should handle error when approving engineer fails", async () => {
    const errorResponse = { response: { data: "Failed to update engineer approval" } };
    apiClient.patch.mockRejectedValueOnce(errorResponse);
    await store.dispatch(approveEngineer({ engineerEmail: "engineer@test.com", approve: true }));
    expect(store.getState().admin.error).toEqual("Failed to update engineer approval");
  });

  test("fetchDeferredTasks should update deferredTasks in state", async () => {
    const mockDeferredTasks = [{ id: 1, status: "deferred" }];
    apiClient.get.mockResolvedValueOnce({ data: mockDeferredTasks });

    await store.dispatch(fetchDeferredTasks());
    expect(store.getState().admin.deferredTasks).toEqual(mockDeferredTasks);
  });

  test("fetchDeferredTasks should handle error when fetching deferred tasks fails", async () => {
    const errorResponse = { response: { data: "Failed to fetch deferred tasks" } };
    apiClient.get.mockRejectedValueOnce(errorResponse);
    await store.dispatch(fetchDeferredTasks());
    expect(store.getState().admin.error).toEqual("Failed to fetch deferred tasks");
  });

  test("fetchAvailableEngineers should update availableEngineers in state", async () => {
    const mockEngineers = [{ id: 1, name: "Engineer 1" }];
    apiClient.get.mockResolvedValueOnce({ data: { engineers: mockEngineers } });

    await store.dispatch(fetchAvailableEngineers());
    expect(store.getState().admin.availableEngineers).toEqual(mockEngineers);
  });

  test("fetchAvailableEngineers should handle error when fetching available engineers fails", async () => {
    const error = new Error("Failed to fetch available engineers");
    apiClient.get.mockRejectedValueOnce(error);
    await store.dispatch(fetchAvailableEngineers());
    expect(store.getState().admin.error).toEqual("Failed to fetch available engineers");
  });

  test("reassignTicket should reassign a ticket successfully", async () => {
    const mockResponse = { success: true };
    apiClient.patch.mockResolvedValueOnce({ data: mockResponse });

    await store.dispatch(reassignTicket({ ticketId: 1, engineerId: 2 }));
    expect(apiClient.patch).toHaveBeenCalledWith(
      "/api/reassign/1/2",
      {},
      { headers: { "Content-Type": "application/json" } }
    );
  });

  test("reassignTicket should handle error when reassigning ticket fails", async () => {
    const error = new Error("Failed to reassign ticket");
    apiClient.patch.mockRejectedValueOnce(error);
    
    await expect(store.dispatch(reassignTicket({ ticketId: 1, engineerId: 2 }))).rejects.toThrow("Failed to reassign ticket");
    expect(apiClient.patch).toHaveBeenCalledWith(
      "/api/reassign/1/2",
      {},
      { headers: { "Content-Type": "application/json" } }
    );
  });

  test("fetchEngineerTasks should update engineerTasks in state", async () => {
    const mockTasks = [{ id: 1, title: "Task for Engineer" }];
    apiClient.get.mockResolvedValueOnce({ data: mockTasks });

    await store.dispatch(fetchEngineerTasks("engineer@test.com"));
    expect(store.getState().admin.engineerTasks).toEqual(mockTasks);
  });

  test("fetchEngineerTasks should handle error when fetching engineer tasks fails", async () => {
    const errorResponse = { response: { data: "Error fetching tasks" } };
    apiClient.get.mockRejectedValueOnce(errorResponse);
    await store.dispatch(fetchEngineerTasks("engineer@test.com"));
    expect(store.getState().admin.error).toEqual("Error fetching tasks");
  });
});