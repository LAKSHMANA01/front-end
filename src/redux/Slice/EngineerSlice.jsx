import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

// AsyncThunk for fetching engineer tasks
export const fetchEngineerTasks = createAsyncThunk(
  'engineer/fetchEngineerTasks',
  async (userId, { rejectWithValue }) => {
    console.log(`userId inside fetchEngineerTasks: ${userId}`);
    try {
      const response = await axios.get(`https://localhost:8000/api/tasks/engineer/engineer3@example.com`);
      console.log(`response.data inside fetchEngineerTasks: ${response.data}`);
      return response.data;
      
    } catch (error) {
      return rejectWithValue(error.response?.data || 'Failed to fetch engineer tasks');
    }
  }
);

// AsyncThunk for updating engineer profile
export const fetchUpdateEngineerProfile = createAsyncThunk(
  'engineer/fetchUpdateEngineerProfile',
  async (updatedData, { rejectWithValue }) => {
    console.log(`updatedData inside fetchUpdateEngineerProfile: ${updatedData}`);
    try {
      const response = await axios.patch(
        `https://localhost:8000/api/updateProfile/engineer/3`, 
        updatedData, 
        {
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );
      console.log(`response.data inside fetchUpdateEngineerProfile: ${response.data}`);
      return response.data;
    } catch (error) {
      return rejectWithValue(error.response?.data || 'Failed to update engineer profile');
    }
  }
);

export const updateTaskStatus = createAsyncThunk(
  'engineer/updateTaskStatus',
  async ({ taskId, status }, { rejectWithValue }) => {
    try {
      const response = await axios.patch(`https://localhost:8000/api/tasks/updateTicketStatus/${taskId}/status/${status}`);
      return { response };  // Return updated task info
    } catch (error) {
      return rejectWithValue(error.response?.data || error.message);
    }
  }
);

export const HazardsTickets = createAsyncThunk(
   'engineer/fetchHazardsTickets',
  async ({ rejectWithValue }) => {
    try {
      console.log("data comeing in axio")
      const response =  await axios.get(`https://localhost:8000/api/hazards/getAllHazards`);
      console.log("response.data inside fetchEngineerTasks:",response.data);
      return response.data; // Return updated task info
    } catch (error) {
      return rejectWithValue(error.response?.data || error.message);
    }
  }
);

// AsyncThunk for Harzards engineer updateing
export const HazardsUpdateTickets = createAsyncThunk(
  'engineer/HazardsUpdateTickets',
  async (updatedData, { rejectWithValue }) => {
    console.log("updatedData inside fetchUpdateHazardsUpdateTickets:", updatedData);
    try {
      const response = await axios.patch(
        `https://localhost:8000/api/hazards/updateHazard/${updatedData._id}`, 
        updatedData, 
        {
          headers: {
            'Content-Type': 'application/json'
          }
        }
      );
      const res = await axios.get(`https://localhost:8000/api/hazards/getAllHazards`);
      console.log("response.data inside  Hazards :", response.data);
      return res.data;
    } catch (error) {
      return rejectWithValue(error.response?.data || 'Failed to HazardsUpdateTickets');
    }
  }
);
// AsyncThunk for Harzards engineer updateing
export const HazardsDeleteTickets = createAsyncThunk(
  'engineer/HazardsDeleteTickets',
  async (updatedData, { rejectWithValue }) => {
    console.log("deletedData inside fetchUpdateHazardsUpdateTickets:", updatedData);
    try {
      const response = await axios.delete(
        `https://localhost:8000/api/hazards/deleteHazard/${updatedData}`, 
       
      );
      const res = await axios.get(`https://localhost:8000/api/hazards/getAllHazards`);
      console.log("response.data inside  Hazards :", response.data);
      return res.data;
    } catch (error) {
      return rejectWithValue(error.response?.data || 'Failed to HazardsUpdateTickets');
    }
  }
);


const engineerSlice = createSlice({
  name: 'engineer',
  initialState: {
    tasks: [],
    updateProfile: [],
    Hazards:[],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      // For fetchEngineerTasks
      .addCase(fetchEngineerTasks.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchEngineerTasks.fulfilled, (state, action) => {
        state.tasks = action.payload;
        state.loading = false;
      })
      .addCase(fetchEngineerTasks.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      
      // For fetchUpdateEngineerProfile
      .addCase(fetchUpdateEngineerProfile.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchUpdateEngineerProfile.fulfilled, (state, action) => {
        state.updateProfile = action.payload;
        state.loading = false;
      })
      .addCase(fetchUpdateEngineerProfile.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      .addCase(updateTaskStatus.fulfilled, (state, action) => {
        const { taskId, status } = action.payload;
        const taskIndex = state.tasks.findIndex(task => task._id === taskId);
        if (taskIndex !== -1) {
          state.tasks[taskIndex].status = status;  // Update status locally
        }
      })
      .addCase(updateTaskStatus.rejected, (state, action) => {
        state.error = action.payload;
      })
      // Hazards data setting here 
      .addCase( HazardsTickets.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase( HazardsTickets.fulfilled, (state, action) => {
        state.Hazards = action.payload;
        console.log("data comeing in fulfilled")
        state.loading = false;
      })
      .addCase( HazardsTickets.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
      // Hazards  update dat patch to engineer
      .addCase( HazardsUpdateTickets.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(HazardsUpdateTickets.fulfilled, (state, action) => {
        state.Hazards = action.payload;
        console.log("data comeing inside udpate Hazards patch")
        state.loading = false;
      })
      .addCase( HazardsUpdateTickets.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })

     // Hazards delete tasks
      .addCase( HazardsDeleteTickets.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(HazardsDeleteTickets.fulfilled, (state, action) => {
        state.Hazards = action.payload;
         
        
        console.log("data comeing inside deleting data delete")
        state.loading = false;
      })
      .addCase( HazardsDeleteTickets.rejected, (state, action) => {
        state.error = action.payload;
        state.loading = false;
      })
  },
});

export default engineerSlice.reducer;
