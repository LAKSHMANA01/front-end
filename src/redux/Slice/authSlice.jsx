import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  user: {
    email: null,
    role: null,
  },
  token: null,
  isAuthenticated: false,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    setUser: (state, action) => {
      state.user = action.payload;
      state.isAuthenticated = true;
    },
    logout: (state) => {
      state.user = { email: null, role: null };
      state.token = null;
      state.isAuthenticated = false;
    }
  }
});

export const { setUser, logout } = authSlice.actions;
export default authSlice.reducer;
