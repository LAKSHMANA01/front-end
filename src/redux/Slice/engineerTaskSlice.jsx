import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

// AsyncThunk for fetching engineer tasks
export const fetchEngineerTasks = createAsyncThunk(
  'engineerTasks/fetchEngineerTasks',
  async ( userId ,{ rejectWithValue }) => {
    console.log("userid from asyncthunk")
    try {                   

      const response = await axios(`https://localhost:8000/api/tickets/engineer/${userId}`);
      console.log("userid  data", response.data)
      return response.data;

    } catch (error) {
      // Return a rejected value to handle errors
      if (!error.response) {
        throw error;
      }
      return rejectWithValue(error.response.data);
    }
  }
);

const engineerTaskSlice = createSlice({
  name: 'engineerTasks',
  initialState: {
    tasks: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchEngineerTasks.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchEngineerTasks.fulfilled, (state, action) => {
        state.tasks = action.payload;
        state.loading = false;
      })
      .addCase(fetchEngineerTasks.rejected, (state, action) => {
        state.error = action.payload || action.error.message;
        state.loading = false;
      });
  },
});

export default engineerTaskSlice.reducer;
