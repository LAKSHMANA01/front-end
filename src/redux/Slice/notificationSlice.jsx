import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

// Fetch notifications (Async Action)
export const fetchNotifications = createAsyncThunk(
  "notifications/fetchNotifications",
  async (userId, { rejectWithValue }) => {
    try {
      console.log("data nofification", userId)
      const response = await axios.get(`https://localhost:8003/api/notifications/getNotifications/marina@gmail.com`);
      console.log("Here Fetchnotification here",response)
      return response.data.notifications.filter((notif) => !notif.isRead);
    } catch (error) {
      return rejectWithValue(error.response.data);
    }
  }
);

// Mark notification as read (Async Action)

export const markAsRead = createAsyncThunk(
  "notifications/markAsRead",
  async (notificationId, { rejectWithValue }) => {
    try {
      console.log("notification marked as read");
      await axios.patch(`https://localhost:8003/api/notifications/updateNotification/67b8114bfa70ebf5e966cb18`, { isRead: true } );
      return "marina@gmail.com" ; // Return the ID to update state
    } catch (error) {
      return rejectWithValue(error.response.data);
    }
  }
);

// Send notification (Async Action)
export const sendNotification = createAsyncThunk(
  "notifications/sendNotification",
  async ({ email, messageToSend, isRead }, { rejectWithValue }) => {
    try {
      const notificationData = { email, message: messageToSend, isRead };
      await axios.post("https://localhost:8003/api/notifications/addNotification", notificationData);
      
    } catch (error) {

      return rejectWithValue(error.response.data);
    }
  }
);

// Notification Slice
const notificationSlice = createSlice({
  name: "notifications", // name of the notification
  initialState: {
    notifications: [],
    loading: false,
    error: null,
  },
  reducers: {},

  extraReducers: (builder) => {
    builder
      // Handle fetchNotifications
      .addCase(fetchNotifications.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchNotifications.fulfilled, (state, action) => {
        state.loading = false;
        state.notifications = action.payload;
      })
      .addCase(fetchNotifications.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })

      // Handle markAsRead
      .addCase(markAsRead.fulfilled, (state, action) => {
        state.notifications = state.notifications.filter(
          (notif) => notif._id !== action.payload
        );
      })

      // Handle sendNotification (Optional Success Feedback)
      .addCase(sendNotification.rejected, (state, action) => {
        state.error = action.payload;
      });
  },
});

// Export actions and reducer
export default notificationSlice.reducer;
