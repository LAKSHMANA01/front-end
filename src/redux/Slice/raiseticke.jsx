import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
export const submitTicket = createAsyncThunk(
  'tickets/submitTicket',
  async (ticketData, { rejectWithValue }) => {
    try {
      console.log("inside data")
      const response = await axios.post(
        ' https://localhost:8000/api/users/raiseTicket/2',
        ticketData,
        {
          headers: {
            'Content-Type': 'application/json',
          },
        }
      );
      console.log("Response data:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error submitting ticket:", error);
      return rejectWithValue(
        error.response?.data || 'Failed to submit ticket'
      );
    }
  }
);

// Redux slice for tickets
const ticketSlice = createSlice({
  name: "tickets",
  initialState: {
    isLoading: false,
    data: [],
    isError: false,
    errorMessage: "",
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(submitTicket.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(submitTicket.fulfilled, (state, action) => {
        state.isLoading = false;
        state.data.push(action.payload); // Add the ticket to the data array
      })
      .addCase(submitTicket.rejected, (state, action) => {
        state.isLoading = false;
        state.isError = true;
        state.errorMessage = action.payload;
      });
  },
});

export default ticketSlice.reducer;
