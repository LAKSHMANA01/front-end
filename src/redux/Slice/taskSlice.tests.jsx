import { configureStore } from '@reduxjs/toolkit';
import taskReducer, { fetchTasks, clearError, clearTasks } from './taskSlice';
import apiClient from '../../utils/apiClient';

// Mock the apiClient
jest.mock('../../utils/apiClient', () => ({
  get: jest.fn()
}));

describe('taskSlice', () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        tasks: taskReducer
      }
    });
    // Clear all mocks before each test
    jest.clearAllMocks();
  });

  describe('fetchTasks', () => {
    test('should handle initial state', () => {
      const state = store.getState().tasks;
      expect(state.tasks).toEqual([]);
      expect(state.loading).toBe(false);
      expect(state.error).toBe(null);
    });

    test('should handle pending state', () => {
      store.dispatch(fetchTasks());
      const state = store.getState().tasks;
      expect(state.loading).toBe(true);
      expect(state.error).toBe(null);
    });

    test('should handle successful fetch', async () => {
      const mockTasks = [
        { id: 1, title: 'Task 1' },
        { id: 2, title: 'Task 2' }
      ];
      
      apiClient.get.mockResolvedValueOnce({ data: mockTasks });

      await store.dispatch(fetchTasks());
      
      const state = store.getState().tasks;
      expect(apiClient.get).toHaveBeenCalledWith('/admin/tasks');
      expect(state.tasks).toEqual(mockTasks);
      expect(state.loading).toBe(false);
      expect(state.error).toBe(null);
    });

    test('should handle fetch error', async () => {
      const errorMessage = 'Failed to fetch tasks';
      apiClient.get.mockRejectedValueOnce({
        response: { data: errorMessage }
      });

      await store.dispatch(fetchTasks());
      
      const state = store.getState().tasks;
      expect(state.loading).toBe(false);
      expect(state.error).toBe(errorMessage);
      expect(state.tasks).toEqual([]);
    });

    test('should handle network error', async () => {
      const errorMessage = 'Network Error';
      apiClient.get.mockRejectedValueOnce(new Error(errorMessage));

      await store.dispatch(fetchTasks());
      
      const state = store.getState().tasks;
      expect(state.loading).toBe(false);
      expect(state.error).toBe(errorMessage);
      expect(state.tasks).toEqual([]);
    });
  });

  describe('synchronous actions', () => {
    test('should clear error state', () => {
      // First set an error
      store = configureStore({
        reducer: {
          tasks: taskReducer
        },
        preloadedState: {
          tasks: {
            error: 'Some error',
            loading: false,
            tasks: []
          }
        }
      });

      store.dispatch(clearError());
      const state = store.getState().tasks;
      expect(state.error).toBe(null);
    });

    test('should clear tasks', () => {
      // First set some tasks
      store = configureStore({
        reducer: {
          tasks: taskReducer
        },
        preloadedState: {
          tasks: {
            tasks: [{ id: 1, title: 'Task 1' }],
            loading: false,
            error: null
          }
        }
      });

      store.dispatch(clearTasks());
      const state = store.getState().tasks;
      expect(state.tasks).toEqual([]);
    });
  });

  describe('error handling', () => {
    test('should handle server error with specific message', async () => {
      const errorMessage = 'Internal Server Error';
      apiClient.get.mockRejectedValueOnce({
        response: { data: errorMessage }
      });

      await store.dispatch(fetchTasks());
      
      const state = store.getState().tasks;
      expect(state.error).toBe(errorMessage);
    });

    test('should handle error with no response data', async () => {
      apiClient.get.mockRejectedValueOnce({});

      await store.dispatch(fetchTasks());
      
      const state = store.getState().tasks;
      expect(state.error).toBeDefined();
    });
  });
});