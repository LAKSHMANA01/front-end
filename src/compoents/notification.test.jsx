// Notifications.test.jsx
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Notifications from './notification';

// Mock react-redux hooks
import { useDispatch, useSelector } from 'react-redux';
jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

// Mock the action creators
import { fetchNotifications, markAsRead } from '../redux/Slice/notificationSlice';
jest.mock('../redux/Slice/notificationSlice', () => ({
  fetchNotifications: jest.fn((userId) => ({ type: 'FETCH_NOTIFICATIONS', payload: userId })),
  markAsRead: jest.fn((notifId) => ({ type: 'MARK_AS_READ', payload: notifId })),
}));

describe('Notifications Component', () => {
  let mockDispatch;
  
  beforeEach(() => {
    jest.clearAllMocks();
    mockDispatch = jest.fn();
    useDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    sessionStorage.clear();
  });

  test('dispatches fetchNotifications in useEffect when userId exists', () => {
    sessionStorage.setItem("email", "user@example.com");
    // Provide default state with empty notifications and not loading
    useSelector.mockImplementation(callback =>
      callback({ notifications: { notifications: [], loading: false } })
    );
    render(<Notifications />);
    
    // Verify that fetchNotifications is called with the userId
    expect(fetchNotifications).toHaveBeenCalledWith("user@example.com");
    expect(mockDispatch).toHaveBeenCalledWith(fetchNotifications("user@example.com"));
  });

  test('does not dispatch fetchNotifications if no userId', () => {
    sessionStorage.removeItem("email");
    useSelector.mockImplementation(callback =>
      callback({ notifications: { notifications: [], loading: false } })
    );
    render(<Notifications />);
    
    // fetchNotifications should not be called because no userId is present
    expect(fetchNotifications).not.toHaveBeenCalled();
  });

  test('renders Loading... text when loading is true', () => {
    sessionStorage.setItem("email", "user@example.com");
    useSelector.mockImplementation(callback =>
      callback({ notifications: { notifications: [], loading: true } })
    );
    render(<Notifications />);
    
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  test('renders "No new notifications" when loading is false and notifications are empty', () => {
    sessionStorage.setItem("email", "user@example.com");
    useSelector.mockImplementation(callback =>
      callback({ notifications: { notifications: [], loading: false } })
    );
    render(<Notifications />);
    
    expect(screen.getByText("No new notifications")).toBeInTheDocument();
  });

  test('renders a list of notifications and dispatches markAsRead when a button is clicked', () => {
    sessionStorage.setItem("email", "user@example.com");
    const mockNotifications = [
      { _id: '1', message: 'Notification 1' },
      { _id: '2', message: 'Notification 2' },
    ];
    useSelector.mockImplementation(callback =>
      callback({ notifications: { notifications: mockNotifications, loading: false } })
    );
    render(<Notifications />);
    
    // Verify that each notification message is rendered
    expect(screen.getByText("Notification 1")).toBeInTheDocument();
    expect(screen.getByText("Notification 2")).toBeInTheDocument();
    
    // There should be a "Mark as Read" button for each notification
    const markButtons = screen.getAllByText("Mark as Read");
    expect(markButtons.length).toBe(2);
    
    // Click the first button and verify dispatch of markAsRead with corresponding id
    fireEvent.click(markButtons[0]);
    expect(mockDispatch).toHaveBeenCalledWith(markAsRead('1'));
    
    // Click the second button and verify dispatch of markAsRead with corresponding id
    fireEvent.click(markButtons[1]);
    expect(mockDispatch).toHaveBeenCalledWith(markAsRead('2'));
  });

  test('matches snapshot', () => {
    sessionStorage.setItem("email", "user@example.com");
    useSelector.mockImplementation(callback =>
      callback({ notifications: { notifications: [], loading: false } })
    );
    const { container } = render(<Notifications />);
    expect(container).toMatchSnapshot();
  });
});