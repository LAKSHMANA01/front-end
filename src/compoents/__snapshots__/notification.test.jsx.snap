// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Component matches snapshot 1`] = `
<div>
  <div
    class="absolute left-1/4 top-16 -translate-x-1/2 bg-white shadow-lg 
 rounded-lg p-4 w-72 md:w-80 z-50 border border-gray-200"
  >
    <h3
      class="text-lg font-semibold mb-4 text-gray-800"
    >
      Notifications
    </h3>
    <p
      class="text-gray-500"
    >
      No new notifications
    </p>
  </div>
</div>
`;
